### PR TITLE
Implement `local`, `locales`, and `onLocaleChanged` for the web

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/54669
Fixes https://github.com/flutter/flutter/issues/48886

Here's the new gallery after I changed my Chrome default language to Russian:

gallery-in-russian

### CODE REVIEW

**Comment by yjbanov** [comment]
@slightfoot Can you please OK including your commits in this PR? https://github.com/flutter/engine/pull/18137#issuecomment-623781517 has the details.

**Comment by slightfoot** [comment]
Yes. Your welcome to my commits.

**Comment by yjbanov** [comment]
> Yes. Your welcome to my commits.

Thanks! Looks like the googlebot wants you to follow this small instruction: We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent`. in this pull request.

**Comment by slightfoot** [comment]
@googlebot I consent

File: lib/web_ui/lib/src/engine/window.dart [hunk 1]
```diff
@@ -168,6 +170,66 @@ class EngineWindow extends ui.Window {
     _onLocaleChangedZone = Zone.current;
   }
 
+  /// The locale used when we fails to get the list from the browser.
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/window.dart:173]: typo: "we fail"

File: lib/web_ui/lib/src/engine/window.dart [hunk 2]
```diff
@@ -168,6 +170,66 @@ class EngineWindow extends ui.Window {
...
+      } else {
+        locales.add(ui.Locale(language));
+      }
+    }
+
+    // To make it easier for the app code, let's not leave the locales list
+    // empty. This way there's fewer corner cases for apps to handle.
+    if (locales.isEmpty) {
+      locales.add(_defaultLocale);
+    }
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/window.dart:224]: Suggestion: return early before the for loop.
```dart
if (!js_util.hasProperty(html.window.navigator, 'languages')) {
  return const [_defaultLocale];
}

if (html.window.navigator.languages.isEmpty) {
  return const [_defaultLocale];
}

for (final String language in html.window.navigator.languages) {
  // ...
}
return locales;
```

This is just a personal preference of mine. It makes reading the code easier when edge cases are handled at the beginning.

File: lib/web_ui/lib/src/engine/window.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart:173]: Done.

File: lib/web_ui/lib/src/engine/window.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart:224]: Done. I actually prefer that too, but missed it this time.

**Comment by yjbanov** [comment]
@slightfoot LMK whether this is working for you. I tested by changing the browser language to "ru" and running the new Flutter Gallery (see screenshot in the PR description).

**Pull request merged**